### PR TITLE
[autosubmit] Add webhook handler

### PR DESCRIPTION
**Pull request by KristinBi**
Created the webhook handler to handle the request.

This intends to solve the issue: https://github.com/flutter/flutter/issues/98703

### CODE REVIEW

File: auto_submit/bin/server.dart [hunk 1]
```diff
@@ -9,25 +9,27 @@ import 'package:auto_submit/helpers.dart';
 import 'package:shelf/shelf.dart';
 import 'package:shelf_router/shelf_router.dart';
 
-class AutosubmitServer {
-  AutosubmitServer();
+Future<Response> webhookHandler(Request request) async {
+  final String? reqHeader = jsonEncode(request.headers);
+  print('Header: $reqHeader');
```
Comment (CaseyHillers) [inline auto_submit/bin/server.dart:14]: Import `package:logging` and use logs instead

https://dart-lang.github.io/linter/lints/avoid_print.html

File: auto_submit/bin/server.dart [hunk 2]
```diff
@@ -9,25 +9,27 @@ import 'package:auto_submit/helpers.dart';
 import 'package:shelf/shelf.dart';
 import 'package:shelf_router/shelf_router.dart';
 
-class AutosubmitServer {
-  AutosubmitServer();
+Future<Response> webhookHandler(Request request) async {
+  final String? reqHeader = jsonEncode(request.headers);
```
Comment (CaseyHillers) [inline auto_submit/bin/server.dart:13]: You can directly log request.headers, not need to json encode it then print it

File: auto_submit/bin/server.dart [hunk 3]
```diff
@@ -9,25 +9,27 @@ import 'package:auto_submit/helpers.dart';
...
+  print('Header: $reqHeader');
+  return Response.ok(
+    jsonEncode(<String, String>{}),
```
Comment (CaseyHillers) [inline auto_submit/bin/server.dart:16]: I would have this parrot back the response body to make it more testable, and drop the headers. There isn't much functionality beyond auth that comes from headers.

File: auto_submit/bin/server.dart [hunk 4]
```diff
@@ -9,25 +9,27 @@ import 'package:auto_submit/helpers.dart';
 import 'package:shelf/shelf.dart';
 import 'package:shelf_router/shelf_router.dart';
 
-class AutosubmitServer {
-  AutosubmitServer();
+Future<Response> webhookHandler(Request request) async {
```
Comment (CaseyHillers) [inline auto_submit/bin/server.dart:12]: This should be moved into it's own file under lib/ like `lib/requests/github_webhook.dart`

File: auto_submit/app.yaml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+runtime: custom
```
Comment (keyonghan) [inline auto_submit/app.yaml:1]: Do we need this file?

File: auto_submit/test/server_test.dart [hunk 6]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2021 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:shelf/shelf.dart';
+import 'package:test/test.dart';
```
Comment (keyonghan) [inline auto_submit/test/server_test.dart:6]: Nit: add an extra space.

File: auto_submit/test/server_test.dart [hunk 7]
```diff
@@ -0,0 +1,21 @@
+// Copyright 2021 The Flutter Authors. All rights reserved.
```
Comment (CaseyHillers) [inline auto_submit/test/server_test.dart:1]: This should be moved test/requests/github_webhook_test.dart

File: auto_submit/test/server_test.dart [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import 'package:test/test.dart';
+import 'package:auto_submit/requests/github_webhook.dart';
+
+void main() {
+  Request req = Request('POST', Uri.parse('http://localhost/'),
+      headers: {
+        'header1': 'header value1',
+      },
+      body: "{\"label1\": \"label1cotent\"}");
+  GithubWebhook githubWebhook = GithubWebhook();
```
Comment (CaseyHillers) [inline auto_submit/test/server_test.dart:15]: Avoid adding logic directly in main. It makes it harder to debug and find what code is being run. Adding basic data models is fine

This should either be in test(...); or in a `setUp(...);`

File: auto_submit/lib/requests/github_webhook.dart [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:logging/logging.dart';
+import 'package:shelf/shelf.dart';
+
+final Logger logger = Logger('GithubWebhook');
+
+class GithubWebhook {
```
Comment (CaseyHillers) [inline auto_submit/lib/requests/github_webhook.dart:12]: dartdoc?

File: auto_submit/lib/requests/github_webhook.dart [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  githubWebhook() {}
+
+  Future<Response> webhookHandler(Request request) async {
```
Comment (CaseyHillers) [inline auto_submit/lib/requests/github_webhook.dart:15]: Maybe call this method `post` instead?

File: auto_submit/lib/requests/github_webhook.dart [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+import 'dart:async';
+
+import 'package:logging/logging.dart';
+import 'package:shelf/shelf.dart';
+
+final Logger logger = Logger('GithubWebhook');
+
+class GithubWebhook {
+  githubWebhook() {}
```
Comment (CaseyHillers) [inline auto_submit/lib/requests/github_webhook.dart:13]:
```suggestion
  const GithubWebhook();
```

File: auto_submit/lib/helpers.dart [hunk 12]
```diff
@@ -7,6 +7,9 @@ import 'dart:io';
 
 import 'package:shelf/shelf.dart';
 import 'package:shelf/shelf_io.dart';
+import 'package:logging/logging.dart';
+
+final Logger logger = Logger('helper');
```
Comment (CaseyHillers) [inline auto_submit/lib/helpers.dart:12]: Let's move the loggers into their own file, like `service/log.dart` and instantiate it like `final Logger log = Logger('auto_submit')`

Changing the tag in every file will make it hard to read the logs and understand what was happening for a request

File: auto_submit/lib/auto_submit.dart [hunk 13]
```diff
@@ -0,0 +1,5 @@
+// Copyright 2019 The Flutter Authors. All rights reserved.
```
Comment (keyonghan) [inline auto_submit/lib/auto_submit.dart:1]: Consistent to be 2022.

File: auto_submit/bin/server.dart [hunk 14]
```diff
@@ -18,7 +20,10 @@ Future main() async {
     );
   }
 
-  final router = Router()..get('/', emptyHandler);
+  GithubWebhook githubWebhook = GithubWebhook();
 
+  final router = Router()
+    ..get('/', emptyHandler)
```
Comment (keyonghan) [inline auto_submit/bin/server.dart:26]: We can remove the emptyHandler now.

**Review by keyonghan — APPROVED** [review]
LGTM with test fix: the new 2022 should be added here: https://github.com/flutter/cocoon/blob/main/licenses/check_licenses.dart#L35

File: auto_submit/lib/requests/github_webhook.dart (hunk 9 above)
Comment (CaseyHillers) [reply auto_submit/lib/requests/github_webhook.dart:10]: No, dartdoc refers to adding a class doc here like

```dart
/// Handler for processing GitHub webhooks.
///
/// On events where a label was added to a pull request, ...
```

See https://dart.dev/guides/language/effective-dart/documentation for the general style guide and suggestions on writing effective documentation

Comment (KristinBi) [reply auto_submit/lib/requests/github_webhook.dart:10]: Gotcha, will add it now. Thanks

Comment (CaseyHillers) [reply auto_submit/lib/requests/github_webhook.dart:10]: FYI this would be above line 10 on the `class GithubWebhook`

**Review by CaseyHillers — APPROVED** [review]
LGTM after the nit is resolved with the doc

**Pull request merged**